### ISSUE TITLE
Consider removing `no-callback-literal` rule

### ISSUE THREAD

**Issue by feross** [standard/eslint-plugin-standard#27]
See discussion at: https://github.com/standard/eslint-plugin-standard/issues/12#issuecomment-298227852

**Comment by yibn2008** [comment]
I strongly recommend removing this rule. First of all, the use-cases of callback with literal is much more common than we expect, currently we can only workaround by change callback name. Secondly, as a non-configurable lint library, I think standard should be more careful for adding new rules, especially when the rules are not very common. Since we cannot configure it, we have to accept.

**Comment by feross** [comment]
Thanks for the feedback @yibn2008, considering it for the next version of `standard`

**Comment by LinusU** [comment]
I personally like this rule, the err-first callback-style is very widely used and I think it's nice to have to be explicit when not following it...

I don't have that strong of an opinion though...

**Comment by xjamundx** [comment]
I suggested this rule originally, but realize it has limited value and is susceptible to false positives.

**Comment by LinusU** [comment]
@targumon In your example, the problematic pattern is that you have no way of handling errors from the `askTheServerForSomething` method. If you want to use callbacks (instead of promises), the best practice way is to let the first argument to the callback be an optional error.

```js
const askTheServerForSomething = (callback) => {
  post()
    .then(callback(null, {x: 1}))
    .catch(ex => callback(ex))
}
```

This ties into having self-contained functions that don't modify global state. In your example it seems like some global handler are invoked in the case of an error in the `post()` function. I would suggest instead putting that error handling in the outer part of the code:

```js
askTheServerForSomething((err, res) => {
  if (err) return globalErrorHandler(err)

  // ...
})
```

It's a bit hard to make better recommendations based on the small example you provided, but I'd be happy to take on any follow up questions/discussion ☺️

**Comment by targumon** [comment]
@LinusU Yes, my example was only meant to illustrate the gist of the actual code.
In reality what my `callback` does is dispatching an action (which causes the app to move to a new state - similar to Redux) and inside `catch` **I use it as well**, just for dispatching a different action.

So it's not like I'm ignoring the errors. And as I said before the linter doesn't complain if I use a different name:

```javascript
const doSomething = (cbFunc) => {
  post()
    .then(cbFunc({x: 1}))
}
```

This is a "cousin" of the `handle-callback-err` check which is only activated if the variable name is `err` or `error`:

```javascript
const foo = (er) => { // this passes
  bar()
}
```

Anyway, thanks for your detailed reply - I learned from it! ^_^

**Comment by xjamundx** [comment]
@bookercodes The simplest explanation looks something like this...

```js
doSomething(function(err, data) {
    if (err) return console.warn(err.stack) // undefined
    console.log(data)
})
```

Or even this example, where did you have the error reading the file? Without the stack trace generated by `new Error()` you will never know.

```
readFile(filename, function(err, data) {
    if (err) console.log(err) // error reading file
})
```

My background as of late has been large enterprise apps where enforcing patterns like this helps save a ton of debugging time. We once had an issue like this where our uncaught exception handler was reporting `undefined undefined` for the error message for TWO MONTHS, which is how look it took for us to figure out that a third party lib was throwing strings, which are missing both the `message` and `stack` properties. That's what got me on the kick of linting away that pain.

Of course there are smaller apps and other patterns that wouldn't have these issues. There are also some systems which rely on passing other data types to a callback function `done(false)` in grunt for example.

**Comment by bookercodes** [comment]
That is a very good point. In my case, I was dealing with an AWS Lambda function `callback`, where it is common to pass a string rather than an `Error`. Passing an `Error` works too so I'm going to do that just in case. If it doesn't hurt and migth help, I am down. Thanks for the response, I apprecaite it.

**Comment by chroriginal** [comment]
It is actually a good habit. However it is absolutely unnecessary when working with `async/await` and `try/catch`. In my case, I called  something like this and got the standard error:
```
cost someValidator = function (propertyName, callback, err) {
    return async function (next) {
      try {
        if (!this.isNew && !this.isModified(propertyName)) return true
        const isValid = await callback({ [propertyName]: this[propertyName] }) // standard linting error
        if (!isValid) throw err
        return next()
      } catch (err) {
        throw err
      }
    }
  }
```

**Comment by chroriginal** [comment]
It is actually a good habit. However, unnecessary when using `async/await` and `try/catch`.

**Comment by LinusU** [comment]
@bookercodes we are running our entire infrastructure on Lambda and have this rule enabled, I would absolutely recommend passing errors instead of strings 👍 

Passing the entire error will also give you the entire stack trace in the CloudWatch logs instead of just the error messages ☺️

**Comment by davestewart** [comment]
FYI, this rule doesn't play well with Electron when intercepting requests:

```js
  session.defaultSession.webRequest.onBeforeRequest(filter, (details, callback) => {
    // some code
    callback({ cancel: false, ... })
  })
```

**Comment by vokar** [comment]
So any plans to remove this rule? It's very inconvenient, and [the way of detecting if a function is a callback by its name](https://github.com/standard/eslint-plugin-standard/blob/master/rules/no-callback-literal.js#L59) is rather hackish.
Especially annoying when adding standardjs to an old project.

**Comment by LinusU** [comment]
@vokar I still don't think that we should remove this rule. Would you mind sharing some examples of the code that you are having problem with? Knowing what the problem is will help us to decide if we want to remove this, or tweak it in some way...
